inference-engine/src/transformations/src/transformations/common_optimizations/convert_nms_gather_path_to_unsigned.cpp

```diff
@@ -0,0 +1,152 @@
+            for (const auto& out_node : out_nodes) {
+                auto& out_rt_info = out_node.get_node()->get_rt_info();
+                out_rt_info["NMS_SELECTED_INDICES"] = std::make_shared<ngraph::VariantWrapper<string>>("");
+                res = true;
+            }
+            return res;
+        };
+
+        auto m = std::make_shared<ngraph::pattern::Matcher>(nms_pattern, matcher_name);
+        register_matcher(m, callback);
+    }
+};
+
+NGRAPH_RTTI_DEFINITION(ngraph::pass::InitNMSPath, "InitNMSPath", 0);
+
+
+class ngraph::pass::PropagateNMSPath: public ngraph::pass::MatcherPass {
+public:
+    NGRAPH_RTTI_DECLARATION;
+
+    PropagateNMSPath(){
+        MATCHER_SCOPE(PropagateNMSPath);
+
+        auto node_pattern = pattern::wrap_type<
+                opset8::Squeeze,
+                opset8::Unsqueeze,
+                opset8::Reshape,
+                op::util::BroadcastBase,
+                opset8::StridedSlice,
+                opset8::VariadicSplit,
+                opset8::Concat,
+                opset8::Convert>();
+
+        ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher &m) {
+            auto node = m.get_match_root();
+            auto& rt_info = node->get_rt_info();
+            bool res = false;
+
+            for (const auto& in_node : node->input_values()) {
+                if (in_node.get_node()->get_rt_info().count("NMS_SELECTED_INDICES")) {
+                    rt_info["NMS_SELECTED_INDICES"] = std::make_shared<ngraph::VariantWrapper<string>>("");
+                    res = true;
+                    break;
+                }
+            }
+
+            return res;
+        };
+
+        auto m = std::make_shared<ngraph::pattern::Matcher>(node_pattern, matcher_name);
+        register_matcher(m, callback);
+    }
+};
+
+NGRAPH_RTTI_DEFINITION(ngraph::pass::PropagateNMSPath, "PropagateNMSPath", 0);
+
+class ngraph::pass::UpdateConvertGather: public ngraph::pass::MatcherPass {
+public:
+    NGRAPH_RTTI_DECLARATION;
+
+    UpdateConvertGather(){
+        MATCHER_SCOPE(UpdateConvertGather);
+
+        auto node_pattern = pattern::wrap_type<op::util::GatherBase>();
+
+        ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher &m) {
+            auto gather = dynamic_pointer_cast<op::util::GatherBase>(m.get_match_root());
+            if (!gather)
+                return false;
+
+            auto indices = gather->input_value(1);
+
+            const auto& rt_info = indices.get_node()->get_rt_info();
+            if (!rt_info.count("NMS_SELECTED_INDICES"))
+                return false;
+
+            auto out_type = element::Type_t::u32;
+            if (indices.get_element_type() == element::Type_t::i64)
+                out_type = element::Type_t::u64;
+
+            if (auto existing_convert = dynamic_pointer_cast<opset8::Convert>(indices.get_node_shared_ptr())) {
+                existing_convert->set_convert_element_type(out_type);
```

Comment:
We have two problems here:
1. We need to call validate_and_infer_types for Convert operation.
2. What if this Convert has more than one consumer?

Reply:
1. agreed and done
2. if Convert has more than one consumer then we need to insert an additional one, done this